### PR TITLE
Integration specs and n+1 problems

### PR DESCRIPTION
**Pull request by dicodiaz**
**Implemented features:**
- Solve the N+1 problem when fetching all posts and their comments for a user
  - Before
pre_fix_n+1
  - After
post_fix_n+1
- Use Capybara to write integration tests for each view:

- User index page:
  - I can see the username of all other users
  - I can see the profile picture for each user
  - I can see the number of posts each user has written
  - When I click on a user, I am redirected to that user's show page

- User show page:
  - I can see the user's profile picture
  - I can see the user's username
  - I can see the number of posts the user has written
  - I can see the user's bio
  - I can see the user's 3 most recent posts
  - I can see a button that lets me view all of a user's posts
  - When I click a user's post, it redirects me to that post's show page
  - When I click to see all posts, it redirects me to the user's post's index page

- Post index page:
  - I can see the user's profile picture
  - I can see the user's username
  - I can see the number of posts the user has written
  - I can see a post's title
  - I can see some of the post's body
  - I can see the 5 most recent comments on a post
  - I can see how many comments a post has
  - I can see how many likes a post has
  - When I click on a post, it redirects me to that post's show page

- Post-show page:
  - I can see the post's title
  - I can see who wrote the post
  - I can see how many comments it has
  - I can see how many likes it has
  - I can see the post body.
  - I can see the username of each commentor
  - I can see the comment each commentor left

- Fix broken requests specs
- Fix broken models specs

Teammates:
@gdumani
@blueberry1312

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -13,6 +13,6 @@ def self.update_posts_counter(user)
   end
 
   def five_most_recent_comments
-    Comment.where(post_id: id).order(updated_at: :desc).limit(5).sort_by(&:updated_at)
+    comments.includes(:author).order(updated_at: :desc).limit(5).sort_by(&:updated_at)
   end
 end
```
Comment (Gambit142) [inline app/models/post.rb:18]: ### Although I am highlighting this file, the proposed change should be made in the `Posts` controller for the `index` action:
- [x] To eliminate the N+1 query problem, I would advise you to create a variable for the post like this:
```ruby
@posts = @user.posts.includes(:comments)
```

Then you can use the @posts variable to replace the `@user.posts.sort` statement in the `app/views/posts/index.html.erb` file. You can try to create many posts and comments in the `db/seeds.rb` file and when you navigate to the `Posts#Index` page, you would see the difference in the database query made in the terminal. 👍 💻 

Kindly implement this change. 😎

File: spec/system/posts_spec.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      expect(page).to have_content(@comment.author.name)
+      expect(page).to have_content(@comment2.author.name)
+    end
+
+    it 'displays the comment each commentor left' do
+      expect(page).to have_content(@comment.text)
+      expect(page).to have_content(@comment2.text)
+    end
+  end
+end
```
Comment (Gambit142) [inline spec/system/posts_spec.rb]: - [x] Based on the requirements, each page should have a test file that contains all tests written for that page. This means that you should have separate test files for the `Post#Index` page and the `Post#Show` page and write the tests in them based on the requirement for that page. 👍 Kindly implement this change. 💻

File: spec/system/users_spec.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      click_link(@post.title)
+      expect(page).to have_current_path(user_post_path(@user, @post))
+    end
+
+    it "redirects to the user's post's index page when clicking on 'See all posts'" do
+      click_link('See all posts')
+      expect(page).to have_current_path(user_posts_path(@user))
+    end
+  end
+end
```
Comment (Gambit142) [inline spec/system/users_spec.rb]: - [x] Based on the requirements, each page should have a test file that contains all tests written for that page. This means that you should have separate test files for the `User#Index` page and the `User#Show` page and write the tests in them based on the requirement for that page. 👍

File: spec/system/posts_spec.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    it 'displays how many likes a post has' do
+      expect(page).to have_content("Comments: #{@post.likes.size}")
+    end
+
+    it "redirects to a post's show page when clicking on it" do
+      click_link(@post.title)
+      expect(page).to have_current_path(user_post_path(@user, @post))
+    end
+  end
+
```
Comment (Gambit142) [inline spec/system/posts_spec.rb]: - [x] For the `Post#Index` page, you forgot to write a test that asserts a _**User**_ can see a section for pagination if there are more posts than fit on the view. Kindly implement this as it is one of the requirements for this milestone.  👍 😎

**Review by Gambit142 — CHANGES_REQUESTED** [review]
Hi @dicodiaz @gdumani @blueberry1312 👋, 

### Good job so far! The tests were well written especially the test that asserts you can see the first comments on a post and the test that asserts you can see the user's first 3 posts. 👏 

## Highlights
- All linters checks passed ✔️ 
- Use of `capybara` gem ✔️ 
- Wrote integration tests for the index and show pages for the `User` model ✔️
- Wrote integration tests for the index and show pages for the `Post` model ✔️  

You still need to work on some issues to go to the next project, but you are almost there!

## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

You can also consider the following:

N/A

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

File: app/models/post.rb (hunk 1 above)
Comment (dicodiaz) [reply app/models/post.rb:18]: Hi @Gambit142, thanks for your review! I'll add the change, but if I were to install `bullet` and navigate to the post's, index page, then it throws me this alert:
image
Taking this into account, I think this is an example of eager loading that we should avoid.

Comment (Gambit142) [reply app/models/post.rb:18]: Hi @dicodiaz, please do not add the `bullet` gem. It was stated in your requirements that you should not use the `bullet` gem. The essence of the exercise is to practise removing **_N + 1 queries_** by writing better codes. Please do not use the `bullet` gem. 👍

Comment (dicodiaz) [reply app/models/post.rb:18]: Gotcha, I was just stating what would happen if I did. I will still add the suggested code, thank you very much.

**Review by Qoosim — APPROVED** [review]
#  *Status: Approved* :tada: :heavy_check_mark: 

Hi @dicodiaz, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

_**I approved my boss PR**_ :+1: 

## Highlights
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**